### ISSUE TITLE
Book Mode:  Improvement Request for ngx-extended-pdf-viewer with Large PDFs

### ISSUE THREAD

**Issue by kingkarki** [stephanrauh/ngx-extended-pdf-viewer#2040]
**I am currently facing** an issue with **ngx-extended-pdf-viewer** when loading large-sized PDFs. While the viewer performs well with smaller PDFs, it significantly slows down when dealing with larger files, particularly when attempting to load them in book mode.

I am seeking guidance on two specific aspects:

**Lazy Loading**: Is there a way to implement lazy loading for PDFs to enhance performance, especially when dealing with larger files? I'm looking for a solution that allows the viewer to load the PDF content progressively rather than attempting to load the entire document at once.

**Loading Success Verification**: Additionally, I am interested in knowing how I can programmatically check whether a PDF has been loaded successfully or not. In the case of larger PDFs, I encounter a white screen and errors during the loading process, and having a mechanism to verify successful loading would be beneficial.

Any assistance or recommendations on optimizing the ngx-extended-pdf-viewer for large PDFs would be greatly appreciated. I want to ensure a smoother user experience, particularly when dealing with substantial document sizes.

` <ngx-extended-pdf-viewer
            [src]="pdf_url"
            height="100%"
            pageViewMode="book"
            [showBorders]="false"
          >
          </ngx-extended-pdf-viewer>`

![image](https://github.com/stephanrauh/ngx-extended-pdf-viewer/assets/756162/560bb57c-467d-40f9-9f84-d33c993735f0)
![image](https://github.com/stephanrauh/ngx-extended-pdf-viewer/assets/756162/ae7fd7b2-ac61-44fd-a59c-a49247c8c8a3)

**Comment by stephanrauh** [comment]
Does the screenshot show one of your large documents? What is your definition of large? Does the file have many pages, or does it have many megabytes?

**Comment by stephanrauh** [comment]
Please check if you habe range request activated. If the server supports range requests, that's the default. You can see this in the network tab of your developer tools. If range requests are active, the PDF is loaded in chunks of 64 kb.

The nice thing is that large PDF files are never loaded completely. When you navigate to another page, you'll see that the viewer starts loading additional chunks. In other words, it already implements lazy loading.

<img width="873" alt="image" src="https://github.com/stephanrauh/ngx-extended-pdf-viewer/assets/3045767/f267ff53-ef02-47de-a317-43729b2ccbf7">

**Comment by stephanrauh** [comment]
The PDF viewer also renders lazily. That works well in the standard display modes, and not quite as well in book mode.

The general idea is that only the visible page is rendered. A few surrounding pages are rendered, too, to support smooth scrolling. The other pages are rendered on demand. You can see this in the DOM. Pages that have been rendered have a canvas (and usually one or more additional divs, such as a text layer):
<img width="508" alt="image" src="https://github.com/stephanrauh/ngx-extended-pdf-viewer/assets/3045767/99095954-7a9b-4a25-8016-8017d7292db2">

Pages that have not been rendered are only an empty div. The screenshot doesn't show that as impressive as I'd like it to do, but even so, you can see that in the red bubble, only the upper-most page has a content. The other pages are empty stubs:
![image](https://github.com/stephanrauh/ngx-extended-pdf-viewer/assets/3045767/92760f60-49cc-4b3b-889c-f5e3fdf90e31)

**Comment by stephanrauh** [comment]
BTW, the preloading algorithm in book mode is unreliable, and I still wonder why. If you're interested in debugging it yourself: it's this method: https://github.com/stephanrauh/pdf.js/blob/a6163a2896c8308b10af28067bb26d9a7db31344/web/pdf_viewer.js#L523. When I took the screenshot above, I noticed that even my showcase demo is broken: initially, it only renders the first page. That why paging to the next page shows empty pages until you release the mouse.

**Comment by stephanrauh** [comment]
Finally, if you PDF has many pages - such as 10000+ pages - your only option is using the single-page mode. That's because the viewer renders an empty div for each page, and even modern browser have difficulties managing 10000 divs. In theory, it's possible to fix this, but that'd be a major modification of pdf.js. I don't want to do that because that means merging updates from pdf.js becomes difficult and error-prone. That's one of the things that need to be fixed by Mozilla (https://github.com/mozilla/pdf.js).

**Comment by kingkarki** [comment]
this is my pdf [https://livescores.hkmenstennisopen.com/livescores/uploaded/Booklet-web.pdf](url)
can you please test with this pdf in book mode, my original  pdf size is more then 200

**Comment by stephanrauh** [comment]
I've manage to make book mode more stable. Along the way I learned that the viewer always loads every page (without rendering it). A weird bug causes the viewer to load the file both with a streaming request and with range requests on a slow network.

If you want to avoid this, disable streaming. You can also experiment with bigger range request chunks:
```ts
  constructor() {
    pdfDefaultOptions.disableRange = false;
    pdfDefaultOptions.disableStream = true;
    pdfDefaultOptions.rangeChunkSize = 256*1024;
   }
```

However, while this reduces network load, it seems to slow down the PDF viewer.

**Comment by stephanrauh** [comment]
The improved version has landed with version 19.0.1. I hope it solves your issue.

Enjoy, and have a happy new year 2024!
Stephan

**Comment by kingkarki** [comment]
i tried this  but not work is it possible to add loading spinner  or somthing when it's loading 
![image](https://github.com/stephanrauh/ngx-extended-pdf-viewer/assets/756162/b342a90f-2ee9-40f4-b019-e347a87b9657)

![image](https://github.com/stephanrauh/ngx-extended-pdf-viewer/assets/756162/d9279c68-53f0-46de-b550-e9c9fbc29ab3)

 ```
 constructor() {
    pdfDefaultOptions.disableRange = false;
    pdfDefaultOptions.disableStream = true;
    pdfDefaultOptions.rangeChunkSize = 256*1024;
   }
```
